### ISSUE TITLE
MCMC Samples with n_changepoints=0

### ISSUE THREAD

**Issue by MaxBirdChemEng** [facebook/prophet#1189]
I am trying to model my time series using the full Bayesian sampling method, with MCMC sampling. I am assuming the trend component will be linear over the entire data range and so I have set growth='linear' and n_changepoints = 0 to reflect this. Here is an example model I tried which fails:

```
m = Prophet(yearly_seasonality=True, daily_seasonality=True, weekly_seasonality=False, 
                      growth='linear', mcmc_samples=100, n_changepoints=0)
```
When I try to predict using this model I get the below traceback error, failed to match shapes (200,) and (1584,). By playing with the code a little I found that the first shape is always double the number of mcmc_samples chosen, and the second shape is the length of the testing data used to predict on. I have found two ways to prevent this error and allow the code to complete but neither make sense to me:

- Remove n_changepoints=0 from the model definition
- Choose a number of mcmc_samples equal to half the length of the testing data. E.g. for 1000 test data points choose mcmc_samples=500

Does anyone know what could be causing this error? I have added a sample of the code along with some training and testing data in the first comment. 

```
<ipython-input-7-9cf46309aba2> in training_prophet(store_list, train_dates, test_dates, MCMC_mode)
    482         ## fit and train model
    483         m.fit(train_data)
--> 484         forecast = m.predict(test_data)

/usr/local/lib/python3.6/dist-packages/fbprophet/forecaster.py in predict(self, df)
   1174         df['trend'] = self.predict_trend(df)
   1175         seasonal_components = self.predict_seasonal_components(df)
-> 1176         intervals = self.predict_uncertainty(df)
   1177 
   1178         # Drop columns except ds, cap, floor, and trend

/usr/local/lib/python3.6/dist-packages/fbprophet/forecaster.py in predict_uncertainty(self, df)
   1376         Dataframe with uncertainty intervals.
   1377         """
-> 1378         sim_values = self.sample_posterior_predictive(df)
   1379 
   1380         lower_p = 100 * (1.0 - self.interval_width) / 2

/usr/local/lib/python3.6/dist-packages/fbprophet/forecaster.py in sample_posterior_predictive(self, df)
   1340                     iteration=i,
   1341                     s_a=component_cols['additive_terms'],
-> 1342                     s_m=component_cols['multiplicative_terms'],
   1343                 )
   1344                 for key in sim_values:

/usr/local/lib/python3.6/dist-packages/fbprophet/forecaster.py in sample_model(self, df, seasonal_features, iteration, s_a, s_m)
   1405         Dataframe with trend and yhat, each like df['t'].
   1406         """
-> 1407         trend = self.sample_predictive_trend(df, iteration)
   1408 
   1409         beta = self.params['beta'][iteration]

/usr/local/lib/python3.6/dist-packages/fbprophet/forecaster.py in sample_predictive_trend(self, df, iteration)
   1462 
   1463         if self.growth == 'linear':
-> 1464             trend = self.piecewise_linear(t, deltas, k, m, changepoint_ts)
   1465         else:
   1466             cap = df['cap_scaled']

/usr/local/lib/python3.6/dist-packages/fbprophet/forecaster.py in piecewise_linear(t, deltas, k, m, changepoint_ts)
   1209         gammas = -changepoint_ts * deltas
   1210         # Get cumulative slope and intercept at each t
-> 1211         k_t = k * np.ones_like(t)
   1212         m_t = m * np.ones_like(t)
   1213         for s, t_s in enumerate(changepoint_ts):

ValueError: operands could not be broadcast together with shapes (200,) (1584,) 

```

**Comment by MaxBirdChemEng** [comment]
Here is a sample of the code, on google colab, as well as a zip folder with a sample set of training and testing data.

[Colab code](https://colab.research.google.com/drive/166ylfvhDbS7EIcIojqfx3ytvfOAKT3jw)
[data.zip](https://github.com/facebook/prophet/files/3755069/data.zip)

**Comment by Igevorse** [comment]
Hi @MaxBirdChemEng!

> By playing with the code a little I found that the first shape is always double the number of mcmc_samples chosen

[By default](https://pystan.readthedocs.io/en/latest/api.html#pystan.stan) PyStan runs 4 chains (4 simulations), each one has `mcmc_samples` samples. Since you specified `mcmc_samples=100`, it is 4 * 100 = 400 samples in total.
Also by default there are `mcmc_samples//2` warmup (burn-in) samples that are discarded, so finally you will get 400 // 2 = 200 samples, which is your initial  `mcmc_samples` doubled.

From my experiments it seems like setting `n_changepoints=0` is unsupported for now (see also https://github.com/facebook/prophet/issues/614#issuecomment-405726942).

Can't wait to hear from @bletham about that.

**Comment by MaxBirdChemEng** [comment]
Hi @Igevorse, thanks for your comments on the mcmc_samples, that makes sense now. 

Hopefully @bletham can comment more on the n_changepoints.

**Comment by bletham** [comment]
Thanks for raising this, this is a bug.

n_changepoints=0 is meant to work (what isn't supported from #614 is to force the trend to have slope=0; with n_changepoints=0 you get a linear trend where the slope `k` and offset `m` are fit).

Because the Stan code expects to fit some non-zero number of changepoint parameters (change magnitudes `delta`), the way n_changepoint=0 is handled is that it actually puts a dummy changepoint at the very first data point in the history:
https://github.com/facebook/prophet/blob/a3ae4715bbc0a18cde35f7b199a671f53df80168/python/fbprophet/forecaster.py#L378-L382

So if the slope is `k` and the amount of change at the changepoint is `delta`, then basically we will be fitting a line with slope `k+delta` everywhere, which is equivalent to having no changepoint at all.

After fitting is completed, the parameters are actually changed to replace `k` with `k + delta` and set `delta` equal to 0:
https://github.com/facebook/prophet/blob/a3ae4715bbc0a18cde35f7b199a671f53df80168/python/fbprophet/forecaster.py#L1146-L1150

So then the model has a single changepoint, but the trend doesn't change there.

The bug we're running into here is happening during predict; I suspect something is wrong with the shape of the parameter array when we do MCMC with n_changepoints = 0. The issue is probably inside `sample_predictive_trend` but I haven't tracked it down entirely yet. Here are a few observations:

* n_changepoints=0 works if not doing MCMC; which is why I think the issue has to do with the shape of the parameter array.
* n_changepoints=0 works with MCMC in the R version. So fundamentally this should work and it's just a bug in the Py version.
* If you set a single changepoint at the first point in your data, then it works:
```
m = Prophet(changepoints=[df['ds'].min()], mcmc_samples=300)
```
and the result is equivalent to n_changepoints=0. So this an easy workaround for now, but we'll definitely have to track down the source of the error and fix the bug.

**Comment by bletham** [comment]
The fix is pushed to PyPI in v0.6
